src/core/sequence_batch_scheduler.cc

```diff
@@ -149,6 +165,101 @@ SequenceBatchScheduler::Create(
   return Status::Success;
 }
 
+Status
+SequenceBatchScheduler::GenerateInitialStateData(
+    const inference::ModelSequenceBatching_InitialState& initial_state,
+    const inference::ModelSequenceBatching_State& state, TritonModel* model)
+{
+  if (initial_state.data_type() != state.data_type()) {
+    return Status(
+        Status::Code::INVALID_ARG,
+        std::string("The data type used for 'initial_state' field of state '") +
+            state.input_name() + "' does not match the state data type.");
+  }
+
+  if (initial_state.name().size() == 0) {
+    return Status(
+        Status::Code::INVALID_ARG,
+        std::string("Field 'name' must be set when using initial_state for "
+                    "state input '") +
+            state.input_name() + "'.");
+  }
+
+  auto initial_state_itr = initial_state_.find(state.input_name());
+  if (initial_state_itr != initial_state_.end()) {
+    return Status(
+        Status::Code::INVALID_ARG, std::string("State input name '") +
+                                       state.input_name() +
+                                       "' specified more than once.");
+  }
+
+  // Check the dimensions to make sure it doesn't have variable-sized dims.
+  for (auto& dim : initial_state.dims()) {
+    if (dim == -1) {
+      return Status(
+          Status::Code::INVALID_ARG,
+          std::string("'initial_state' field for state input name '") +
+              state.input_name() + "' contains variable dimensions.");
+    }
+  }
+
+  const auto& initial_state_pair = initial_state_.emplace(
+      std::piecewise_construct, std::forward_as_tuple(state.input_name()),
+      std::forward_as_tuple(initial_state.name()));
+  auto& initial_state_data = initial_state_pair.first->second;
+
+  // Calculate total memory byte size
+  auto element_count = GetElementCount(initial_state.dims());
+  size_t dtype_byte_size = GetDataTypeByteSize(initial_state.data_type());
+  size_t total_byte_size = element_count * dtype_byte_size;
+
+  // Custom handling for TYPE_BYTES
+  if (dtype_byte_size == 0) {
+    total_byte_size = sizeof(int32_t) * element_count;
+  }
+
+  switch (initial_state.state_data_case()) {
+    case inference::ModelSequenceBatching_InitialState::StateDataCase::
+        kZeroData: {
+      initial_state_data.data_ = std::make_shared<AllocatedMemory>(
+          total_byte_size, TRITONSERVER_MEMORY_CPU /* memory_type */,
+          0 /* memory_type_id */);
+      break;
+    }
+    case inference::ModelSequenceBatching_InitialState::StateDataCase::
+        kDataFile: {
+      initial_state_data.file_data_.reset(new std::string());
```

Comment:
I don't think you need to store the file data explicitly. Instead of making `data_` a reference to `file_data_`, you can allocate memory of file size in `data_` and copy the file data to `data_`. There is extra copy but it is at load time so it is probably okay, and this drop the extra `file_data_` variable.

Reply:
Yes, exactly. I wanted to avoid the data_file copy. Ok, I'll change this accordingly.